### PR TITLE
feat: add deno tasks sidebar

### LINKED ISSUE #852: Deno Tasks Sidebar (NPM Scripts)
**Is your feature request related to a problem? Please describe.**

Node.js projects have an "NPM Scripts" section in the VSCode Explorer sidebar with buttons that allow us to quickly run scripts contained in package.json. I usually use it to quickly lint the project, run the dev server or run the tests. 

**Describe the solution you'd like**

It would be really nice for Deno extension to add the same feature, maybe called "Deno Tasks". This should only show when there is a deno.json/c file present in the project / in a directory in the project. There should be a run button and possibly a debug button. If you double click a task it takes you to the definition in deno.json/c file.

NPM Scripts also allow for monorepos where each package.json found in the project creates a nested section in the sidebar area.

Finally NPM Scripts include an "npm install" button at the bottom of the list of scripts. I propose a "deno cache" button which does a similar thing.

image

### PR DESCRIPTION
**Pull request by ArmaanAS**
(Closes #852) This pull request adds a Deno Tasks menu to the Explorer menu, which is the equivalent to the NPM Script menu built in to VSCode. I've tried to match the features and style by using the [original code as reference.](https://github.com/microsoft/vscode/blob/main/extensions/npm/src/npmView.ts) It's MIT licensed and I've modified it heavily so hopefully that shouldn't be an issue.

<p align="center">
  Deno Tasks vs NPM Scripts
  <br>
  <i>Deno Tasks compared with NPM Scripts.</i>
</p>

## Features

- Display the list of Deno tasks defined in the deno.json[c] config file.
- Run, debug or open the source for each deno task.
- Refresh the menu, but it will refresh automatically when a deno.json[c] file changes / gets renamed or when the `"deno.defaultTaskCommand"` workspace config changes.
- Only displayed when the LSP is enabled. Otherwise, (for non-Deno projects) this menu will be hidden.

<p align="center">
  Deno Tasks Standalone
  <br>
  <i>Deno Tasks as it's own section in the activity bar.</i>
</p>

## Improvements

I've only minimally used the Deno LSP as I'm not very familiar with it but it seems like some of the processing should be done by the LSP, for example:
1. Getting the exact filename of the config file (`deno.json` vs `deno.jsonc`).
2. Getting the task source line number / character in line.

The LSP is only being used to retrieve a list of task names / commands.

### CODE REVIEW

**Comment by ArmaanAS** [comment]
I appreciate that this is a big pull request, and just a fairly complex feature. If there's anything you would like me to break down further, I'll happily try my best. If there's anything that I should add or remove or modify, let me know. I'm not sure if it's standard to give a description of each file changed but I'll give an overview:

### `client/package.json`
- installed jsonc-parser npm library to get the line number and character in line for each deno task from `deno.json[c]`.

### `package.json`
- Defined the UI view.
- Defined 4 new commands, `deno.task.run`, `deno.task.debug`, `deno.task.definition.open` and `deno.tasks.refresh`.
- Disabled the new commands from displaying in the command palette.
- Defined a workspace configuration variable `deno.defaultTaskCommand` which is either `"open"` or `"run"`.
- Defined UI buttons in the new view, inline run button, inline debug button, refresh button in title.

### `client/src/extension.ts`
- Register the tree view with a `DenoTasksTreeDataProvider` as the `treeDataProvider`.
- Add a workspace configuration variable change listener which refreshes the sidebar when `deno.defaultTaskCommand` changes.
- Add a save file listener which refreshes the sidebar when a `deno.json[c]` file changes.
- Add a rename file listener which refreshes the sidebar when a `deno.json[c]` file changes.

### `client/src/util.ts`
- Moved `isWorkspaceFolder()` function from client/src/tasks.ts to here.
- Add a `readTaskDefinitions()` function which reads a `deno.json[c]` file and retrieves all the tasks' line numbers and position in line. This uses the jsonc-parser library.

### `client/src/tasks_sidebar.ts`
Simply put this contains all the data types and classes needed to get the `TreeDataProvider` to work. Each `deno.json[c]` and task are items in the tree. Each item can have various properties and function associated with them. This file registers the 4 commands. I've copied the DenoTask class from `client/src/tasks.ts` but modified it quite a bit to get it to work with the rest of the logic.


Hopefully that's enough to help you start to review the code. Thanks.

**Comment by bartlomieju** [comment]
@nayeemrmn please review this PR

File: client/src/tasks_sidebar.ts [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    ));
+    subscriptions.push(commands.registerCommand(
+      "deno.task.definition.open",
+      this.#openTaskDefinition,
+      this,
+    ));
+    subscriptions.push(commands.registerCommand(
+      "deno.tasks.refresh",
+      this.refresh.bind(this),
+    ));
```
Comment (nayeemrmn) [inline client/src/tasks_sidebar.ts:237]: I think we should avoid granular namespaces for command names, can we rename these to the following?
- `deno.client.runTask`
- `deno.client.debugTask`
- `deno.client.openTaskDefinition`
- `deno.client.refreshTasks`

File: client/package-lock.json [hunk 2]
```diff
@@ -506,411 +513,5 @@
...
-      "version": "1.0.2",
-      "resolved": "https://registry.npmjs.org/wrappy/-/wrappy-1.0.2.tgz",
-      "integrity": "sha1-tSQ9jz7BqjXxNkYFvA0QNuMKtp8=",
-      "dev": true
-    },
-    "yallist": {
-      "version": "4.0.0",
-      "resolved": "https://registry.npmjs.org/yallist/-/yallist-4.0.0.tgz",
-      "integrity": "sha512-3wdGidZyq5PB084XLES5TpOSRA3wjXAlIWMhum2kRcv/41Sn2emQ0dycQW4uZXLejwKvg6EsvbdlVL+FYEct7A=="
-    }
```
Comment (nayeemrmn) [inline client/package-lock.json:914]: Do you know what the lock file removals are? I `git apply`'d this PR and ran `npm install` and don't get them.

Comment (ArmaanAS) [reply client/package-lock.json]: I have no idea why installing jsonc-parser before caused those dependencies to disappear but I've run npm install again and they have reappeared.

File: client/src/tasks_sidebar.ts (hunk 1 above)
Comment (ArmaanAS) [reply client/src/tasks_sidebar.ts]: Ok, I've made these changes.

**Review by nayeemrmn — APPROVED** [review]
LGTM, great feature!

In the near future I'll work on moving the task def acquisition to the LSP.

File: package.json [hunk 3]
```diff
@@ -595,7 +633,80 @@
...
+        },
+        {
+          "command": "deno.client.openTaskDefinition",
+          "when": "false"
+        },
+        {
+          "command": "deno.client.refreshTasks",
+          "when": "false"
+        }
+      ],
```
Comment (nayeemrmn) [inline package.json:668]: @ArmaanAS Does this section do anything considering all the conditions are `false`? They don't show up on the command palette, though `deno.client.refreshTasks` is the only one that should.

Comment (ArmaanAS) [reply package.json:668]: Without these lines added, all of these commands will show up in the command palette but they won’t function properly.

Maybe refreshTasks could be displayed in the command palette, if you think that’s useful.
